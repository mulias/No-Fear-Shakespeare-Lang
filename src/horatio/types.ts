export interface IO {
  print(text: string): void;
  read_char(callback: (input: string) => void): void;
  read_int(callback: (input: string) => void): void;
  debug: boolean;
  printDebug(text: string): void;
  clear(): void;
}
