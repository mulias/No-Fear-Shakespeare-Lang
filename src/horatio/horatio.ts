import Compiler from "./compiler";
import * as Ast from "./ast";
import Program from "./program";

/**
 * Horatio
 * A Javascript compiler for the Shakespeare Programming Language
 *
 * @author Miles Zimmerman
 */
import { IO } from "./types";

declare global {
  interface Window {
    Horatio: typeof Horatio;
  }
}

class Horatio extends Compiler {
  constructor(
    source: string | undefined,
    ast: Ast.Program,
    io: IO,
    program: Program,
  ) {
    super(source, ast, io, program);
  }
}

if (typeof window !== "undefined") {
  window.Horatio = Horatio;
}

export default Horatio;
