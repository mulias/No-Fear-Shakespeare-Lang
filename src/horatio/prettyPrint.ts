import * as Ast from "./ast";
import Formatter, { Part, Subpart, Line } from "./formatter";

export const prettyPrint = (ast: Ast.Program): string => {
  const formatter = new Formatter();
  const { title, declarations, parts } = ast.visit(formatter);

  return [title, declarations.join("\n"), ...parts.map(prettyPrintPart)].join(
    "\n\n",
  );
};

const prettyPrintPart = ({ heading, subparts }: Part): string => {
  return [indent(heading, 20), ...subparts.map(prettyPrintSubpart)].join(
    "\n\n",
  );
};

const prettyPrintSubpart = ({ heading, body }: Subpart): string => {
  return [indent(heading, 20), ...body.map(prettyPrintLine)].join("\n\n");
};

const prettyPrintLine = (line: Line | string): string => {
  if (typeof line === "string") {
    return line;
  } else {
    return [line.name, indent(wrap(line.text, 76), 4)].join("\n");
  }
};

const indent = (text: string, spaces: number): string => {
  const indentation = " ".repeat(spaces);

  return text
    .split("\n")
    .map((line) => indentation + line)
    .join("\n");
};

const wrap = (text: string, maxLength: number): string => {
  const words = text.split(" ");
  let lines: string[] = [];
  let line = "";

  words.forEach((word) => {
    if (line.length + word.length + 1 < maxLength) {
      line = line + " " + word;
    } else {
      lines.push(line);
      line = word;
    }
  });

  lines.push(line);

  return lines.join("\n");
};
