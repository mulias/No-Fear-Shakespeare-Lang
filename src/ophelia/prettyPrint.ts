import * as Ast from "./ast";
import { AssertNeverError } from "../util";

export function prettyPrint(ast: Ast.Program): string {
  const printer = new PrettyPrinter();
  return printer.printProgram(ast);
}

class PrettyPrinter {
  private indentLevel = 0;
  private readonly indentSize = 2;

  private indent(): string {
    return " ".repeat(this.indentLevel * this.indentSize);
  }

  private withIndent<T>(fn: () => T): T {
    this.indentLevel++;
    const result = fn();
    this.indentLevel--;
    return result;
  }

  private printTemplateString(templateString: Ast.TemplateString): string {
    return templateString.value
      .map((segment) => {
        if (segment.type === "template_string_segment") {
          return segment.value;
        } else if (segment.type === "template_var_segment") {
          return `{${segment.value}}`;
        } else {
          throw new Error(
            `Unknown template segment type: ${(segment as any).type}`,
          );
        }
      })
      .join("");
  }

  printProgram(program: Ast.Program): string {
    const parts: string[] = [];

    if (program.title) {
      parts.push(`## title: ${this.printTemplateString(program.title)}`);
    }

    // Add var declarations
    for (const [varName, description] of program.varDeclarations) {
      parts.push(`## var ${varName}: ${this.printTemplateString(description)}`);
    }

    if (program.items.length > 0) {
      // Add empty line before first act if there are var declarations or title
      if (program.title || program.varDeclarations.size > 0) {
        parts.push("");
      }

      // Print each program item
      for (let i = 0; i < program.items.length; i++) {
        const programItem = program.items[i];
        if (programItem) {
          const printedItem = this.printProgramItem(programItem);
          parts.push(printedItem);

          // Add blank line after acts (except the last one)
          if (programItem.type === "act" && i < program.items.length - 1) {
            parts.push("");
          }
        }
      }
    }

    return parts.join("\n");
  }

  private printProgramItem(item: Ast.ProgramItem): string {
    switch (item.type) {
      case "act":
        return this.printAct(item);
      case "comment":
        return this.printComment(item);
      default:
        throw new AssertNeverError(item);
    }
  }

  private printAct(act: Ast.Act): string {
    const parts: string[] = [];

    // Add description doc comment if present
    if (act.description) {
      parts.push(
        `## description: ${this.printTemplateString(act.description)}`,
      );
    }

    const header = `${act.actId} {`;
    const footer = "}";

    if (act.items.length === 0) {
      parts.push(`${header}\n${footer}`);
      return parts.join("\n");
    }

    const items = this.withIndent(() => {
      const itemParts: string[] = [];
      for (let i = 0; i < act.items.length; i++) {
        const item = act.items[i];
        if (item) {
          itemParts.push(this.printActItem(item));

          if (i < act.items.length - 1) {
            itemParts.push("");
          }
        }
      }
      return itemParts.join("\n");
    });

    parts.push(`${header}\n${items}\n${footer}`);
    return parts.join("\n");
  }

  private printActItem(item: Ast.ActItem): string {
    switch (item.type) {
      case "scene":
        return this.printScene(item);
      case "comment":
        return this.printComment(item);
      default:
        throw new AssertNeverError(item);
    }
  }

  private printScene(scene: Ast.Scene): string {
    const parts: string[] = [];

    // Add description doc comment if present
    if (scene.description) {
      parts.push(
        `${this.indent()}## description: ${this.printTemplateString(
          scene.description,
        )}`,
      );
    }

    const header = `${this.indent()}${scene.sceneId} {`;
    const footer = `${this.indent()}}`;

    if (scene.directions.length === 0) {
      parts.push(`${header}\n${footer}`);
      return parts.join("\n");
    }

    const directions = this.withIndent(() => {
      const dirParts: string[] = [];
      for (let i = 0; i < scene.directions.length; i++) {
        const dir = scene.directions[i];
        if (dir) {
          dirParts.push(this.printDirection(dir));

          // Apply spacing rules based on current and next direction types
          if (i < scene.directions.length - 1) {
            const nextDir = scene.directions[i + 1];
            if (nextDir) {
              const currentIsStage = this.isStageDirection(dir);
              const nextIsStage = this.isStageDirection(nextDir);

              // Add a blank line unless both items are stage directions, in
              // which case they should be grouped
              if (!currentIsStage || !nextIsStage) {
                dirParts.push("");
              }
            }
          }
        }
      }
      return dirParts.join("\n");
    });

    parts.push(`${header}\n${directions}\n${footer}`);
    return parts.join("\n");
  }

  private isStageDirection(direction: Ast.Direction): boolean {
    return (
      direction.type === "stage" ||
      direction.type === "unstage" ||
      direction.type === "unstage_all"
    );
  }

  private printDirection(direction: Ast.Direction): string {
    switch (direction.type) {
      case "dialogue":
        return this.printDialogue(direction);
      case "stage":
        return this.printStage(direction);
      case "unstage":
        return this.printUnstage(direction);
      case "unstage_all":
        return `${this.indent()}unstage_all`;
      case "comment":
        return this.printComment(direction);
      default:
        throw new AssertNeverError(direction);
    }
  }

  private printComment(comment: Ast.Comment): string {
    return `${this.indent()}# ${comment.content}`;
  }

  private printDialogue(dialogue: Ast.Dialogue): string {
    const header = `${this.indent()}${dialogue.speakerVarId} {`;
    const footer = `${this.indent()}}`;

    if (dialogue.lines.length === 0) {
      return `${header}\n${footer}`;
    }

    const lines = this.withIndent(() => {
      const lineParts: string[] = [];
      for (let i = 0; i < dialogue.lines.length; i++) {
        const line = dialogue.lines[i];
        if (line) {
          lineParts.push(this.printStatementOrComment(line));

          // Add blank line after statement if it has followedByBlankLine flag
          // and it's not the last statement in the dialogue
          if (i < dialogue.lines.length - 1) {
            if (line.type !== "comment" && line.followedByBlankLine) {
              lineParts.push("");
            }
          }
        }
      }
      return lineParts.join("\n");
    });

    return `${header}\n${lines}\n${footer}`;
  }

  private printStatementOrComment(item: Ast.StatementOrComment): string {
    if (item.type === "comment") {
      return this.printComment(item);
    } else {
      return this.printStatement(item);
    }
  }

  private printStage(stage: Ast.Stage): string {
    const args = stage.varIds.join(", ");
    return `${this.indent()}stage(${args})`;
  }

  private printUnstage(unstage: Ast.Unstage): string {
    const args = unstage.varIds.join(", ");
    return `${this.indent()}unstage(${args})`;
  }

  private printStatement(statement: Ast.Statement): string {
    switch (statement.type) {
      case ".set":
        return `${this.indent()}@you.set(${this.printExpression(
          statement.value,
        )})`;
      case ".print_char":
        return `${this.indent()}@you.print_char`;
      case ".print_int":
        return `${this.indent()}@you.print_int`;
      case ".read_char":
        return `${this.indent()}@you.read_char`;
      case ".read_int":
        return `${this.indent()}@you.read_int`;
      case ".push_self":
        return `${this.indent()}@you.push_self`;
      case ".push_me":
        return `${this.indent()}@you.push_me`;
      case ".pop":
        return `${this.indent()}@you.pop`;
      case "test_eq":
      case "test_gt":
      case "test_lt":
      case "test_not_eq":
      case "test_not_gt":
      case "test_not_lt":
        return `${this.indent()}${statement.type}(${this.printExpression(
          statement.left,
        )}, ${this.printExpression(statement.right)})`;
      case "if":
        const condition = statement.is ? "if_true" : "if_false";
        return `${this.indent()}${condition}(${this.printStatement(
          statement.then,
        ).trim()})`;
      case "goto":
        return `${this.indent()}goto(${statement.labelId})`;
      default:
        throw new AssertNeverError(statement);
    }
  }

  private printExpression(expression: Ast.Expression): string {
    return this.printExpressionWithPrecedence(expression, 0);
  }

  private printExpressionWithPrecedence(
    expression: Ast.Expression,
    parentPrecedence: number,
  ): string {
    switch (expression.type) {
      case "int":
        return expression.value.toString();
      case "char":
        return `'${expression.value}'`;
      case "var":
        return expression.id;
      case "you":
        return "@you";
      case "arithmetic":
        const precedence = this.getOperatorPrecedence(expression.op);
        const left = this.printExpressionWithPrecedence(
          expression.left,
          precedence,
        );
        const right = this.printExpressionWithPrecedence(
          expression.right,
          precedence,
        );
        const result = `${left} ${expression.op} ${right}`;

        // Add parentheses if this operator has lower precedence than parent
        if (precedence < parentPrecedence) {
          return `(${result})`;
        }
        return result;
      case "unary":
        return `${expression.op}(${this.printExpressionWithPrecedence(
          expression.operand,
          0,
        )})`;
      default:
        throw new AssertNeverError(expression);
    }
  }

  private getOperatorPrecedence(op: Ast.ArithmeticOp): number {
    // PEMDAS precedence: higher number = higher precedence
    switch (op) {
      case "*":
      case "/":
      case "%":
        return 2; // Multiplication, Division, Modulo
      case "+":
      case "-":
        return 1; // Addition, Subtraction
      default:
        return 0;
    }
  }
}
