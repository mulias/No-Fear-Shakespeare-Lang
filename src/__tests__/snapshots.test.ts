import { Possum } from "../possum";
import { Ophelia } from "../ophelia";
import { Yorick } from "../yorick";
import { prettyPrint } from "../horatio/prettyPrint";
import Horatio from "../horatio/compiler";
import { IO } from "../horatio/types";

describe("NFSPL to SPL Snapshots", () => {
  // Mock IO interface for testing
  const mockIO: IO = {
    print: () => {},
    read: () => {},
    debug: false,
    printDebug: () => {},
    clear: () => {},
  };

  /**
   * Transpile NFSPL source to SPL and create snapshots
   */
  async function transpileNfspl(source: string): Promise<string> {
    // Step 1: Parse NFSPL with Possum
    const possum = new Possum(source);
    const possumAst = await possum.run();

    // Step 2: Transform with Ophelia
    const ophelia = new Ophelia(possumAst);
    const opheliaAst = ophelia.run();

    // Step 3: Transpile with Yorick
    const yorick = new Yorick(opheliaAst);
    const yorickAst = yorick.run();

    // Step 4: Format to SPL
    return prettyPrint(yorickAst);
  }

  /**
   * Verify that generated SPL can be parsed back
   */
  function validateSpl(splSource: string): void {
    try {
      const horatio = Horatio.fromSource(splSource, mockIO);
      expect(horatio.ast).toBeTruthy();
    } catch (e) {
      throw new Error(`Generated SPL is invalid: ${e}`);
    }
  }

  describe("example programs", () => {
    it("should transpile hi.nfspl to SPL", async () => {
      const nfsplSource = `Main {
  PrintHi {
    stage(a, b)

    b {
      @you.set(72)
      @you.print_char
    }

    b {
      @you.set(@you + 1)
      @you.print_char
    }

    b {
      @you.set(10)
      @you.print_char
    }

    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);

      // Verify the generated SPL is valid by parsing it back
      validateSpl(spl);

      expect(spl).toMatchSnapshot();
    });

    it("should transpile fizzbuzz.nfspl to SPL", async () => {
      const nfsplSource = `## To Fizz, Perchance To Buzz
## declare {n}, a counter of loops
## declare {out}, a poet and orator

Main {
  Start {
    stage(n, out)

    out {
      @you.set(0)
    }
  }

  Loop {
    out { @you.set(@you + 1) }

    out {
      test_gt(@you, 100)
      if_true(goto(End))
    }

    out {
      test_eq(@you % 15, 0)
      if_true(goto(Fizzbuzz))
    }

    out {
      test_eq(@you % 3, 0)
      if_true(goto(Fizz))
    }

    out {
      test_eq(@you % 5, 0)
      if_true(goto(Buzz))
    }

    n {
      @you.set(n)
      @you.print_int
      @you.set(10)
      @you.print_char
      goto(Loop)
    }
  }

  Fizzbuzz {
    n {
      @you.set('f')
      @you.print_char
      @you.set('i')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('b')
      @you.print_char
      @you.set('u')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char

      @you.set(10)
      @you.print_char

      goto(Loop)
    }
  }

  Fizz {
    n {
      @you.set('f')
      @you.print_char
      @you.set('i')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char

      @you.set(10)
      @you.print_char

      goto(Loop)
    }
  }

  Buzz {
    n {
      @you.set('b')
      @you.print_char
      @you.set('u')
      @you.print_char
      @you.set('z')
      @you.print_char
      @you.set('z')
      @you.print_char

      @you.set(10)
      @you.print_char

      goto(Loop)
    }
  }

  End {
    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });

    it("should transpile simplified reverse program", async () => {
      const nfsplSource = `Main {
  Init {
    stage(stack, count)
    stack { @you.set(0) }
  }

  GetInput {
    count {
      @you.read_char
      @you.push_self
    }
    stack {
      @you.set(@you + 1)
      test_gt(@you, 0)
    }
    count {
      if_true(goto(GetInput))
      @you.pop
    }
  }

  PrintReversed {
    count {
      @you.pop
      @you.print_char
    }
    stack {
      @you.set(@you - 1)
      test_gt(@you, 0)
    }
    count { if_true(goto(PrintReversed)) }
  }

  End {
    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });
  });

  describe("language features", () => {
    it("should transpile basic I/O operations", async () => {
      const nfsplSource = `Main {
  IOTest {
    stage(reader, writer)

    reader {
      @you.read_char
      @you.read_int
    }

    writer {
      @you.set('A')
      @you.print_char
      @you.set(42)
      @you.print_int
    }

    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });

    it("should transpile arithmetic operations", async () => {
      const nfsplSource = `Main {
  ArithmeticTest {
    stage(calc, result)

    calc {
      @you.set(10 + 5)
      @you.set(20 - 7)
      @you.set(6 * 7)
      @you.set(15 / 3)
      @you.set(17 % 5)
    }

    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });

    it("should transpile stack operations", async () => {
      const nfsplSource = `Main {
  StackTest {
    stage(pusher, stack)

    pusher {
      @you.set(42)
      @you.push_self
      @you.push_me
    }

    stack {
      @you.pop
      @you.pop
    }

    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });

    it("should transpile control flow", async () => {
      const nfsplSource = `Main {
  ControlTest {
    stage(tester, value)

    tester {
      test_eq(@you, 42)
      if_true(goto(Success))
      test_gt(@you, 0)
      if_false(goto(Failure))
    }
  }

  Success {
    stage(winner, output)
    winner { @you.set('W') }
    unstage_all
  }

  Failure {
    stage(loser, output)
    loser { @you.set('L') }
    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });

    it("should transpile comparison operations", async () => {
      const nfsplSource = `Main {
  ComparisonTest {
    stage(left, right)

    left {
      test_eq(@you, right)
      test_gt(@you, right)
      test_lt(@you, right)
      test_not_eq(@you, right)
      test_not_gt(@you, right)
      test_not_lt(@you, right)
    }

    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });

    it("should transpile variable references", async () => {
      const nfsplSource = `Main {
  VariableTest {
    stage(a, b)

    a {
      @you.set(10)
    }

    b {
      @you.set(a + 5)
      @you.set(a * b)
    }

    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });

    it("should transpile character literals", async () => {
      const nfsplSource = `Main {
  CharTest {
    stage(char_writer, output)

    char_writer {
      @you.set('A')
      @you.set('Z')
      @you.set('0')
      @you.set('9')
    }

    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });

    it("should transpile positive numbers", async () => {
      const nfsplSource = `Main {
  NumberTest {
    stage(numbers, output)

    numbers {
      @you.set(0)
      @you.set(1)
      @you.set(42)
      @you.set(100)
    }

    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });

    it("should transpile multiple acts and scenes", async () => {
      const nfsplSource = `Act1 {
  Scene1 {
    stage(a, b)
    a { @you.set(1) }
    unstage_all
  }

  Scene2 {
    stage(c, d)
    c { @you.set(2) }
    unstage_all
  }
}

Act2 {
  Scene1 {
    stage(e, f)
    e { @you.set(3) }
    unstage_all
  }
}`;

      const spl = await transpileNfspl(nfsplSource);
      expect(spl).toMatchSnapshot();
    });
  });
});
