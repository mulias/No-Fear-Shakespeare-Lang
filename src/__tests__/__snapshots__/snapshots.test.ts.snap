// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NFSPL to SPL Snapshots example programs should transpile fizzbuzz.nfspl to SPL 1`] = `
"Oberon and the healthy wolf.

Oberon, the n variable.
Benvolio, the out variable.

                    Act I: Main.

                    Scene I: Start.

[Enter Oberon and Benvolio]

Benvolio:
     Thou art nothing.

                    Scene II: Loop.

Benvolio:
     You are the sum of you and sky.

Benvolio:
     Is you jollier than sum of sum of bold peaceful nose and healthy reddest
    trustworthy delicious good Heaven and rich sweetest fair embroidered
    charming charming hamster? If so, we shall proceed to scene VI.

Benvolio:
     Is remainder of the quotient between you and sum of sum of sum of nose and
    rural cat and loving white happiness and smooth fair sweet cat as smooth as
    nothing? If so, let us proceed to scene III.

Benvolio:
     Is remainder of the quotient between you and sum of cat and pretty joy as
    old as nothing? If so, We shall proceed to scene IV.

Benvolio:
     Is remainder of the quotient between you and sum of summer's day and good
    warm hero as beautiful as nothing? If so, We shall proceed to scene V.

Oberon:
     You are Oberon. Open your heart! You are the sum of proud joy and pretty
    mighty rich hero. Speak your mind! we shall return to scene II.

                    Scene III: Fizzbuzz.

Oberon:
     You are the sum of sum of sum of loving Lord and honest cunning plum and
    beautiful sunny warm loving noble Heaven and bold smooth happy embroidered
    blossoming cunning sky. Speak thine mind! You are the sum of sum of sum of
    happiness and sweet happy prompt King and cute little cunning fine rural
    summer's day and brave charming delicious tiny trustworthy beautiful purse.
    Speak thine mind! Thou art the sum of sum of sum of sum of little sky and
    handsome reddest beautiful flower and sunny golden happy tiny hamster and
    lovely happy cute brave white angel and trustworthy sweet fair healthy
    blossoming little sky. Speak thine mind! Thou art the sum of sum of sum of
    sum of happy kingdom and amazing charming cunning hamster and healthy noble
    loving sweetest nose and clearest loving delicious warm cute nose and
    embroidered rich rural trustworthy cute amazing summer's day. Speak thy
    mind! Thou art the sum of sum of healthy rose and golden fair healthy
    delicious reddest happiness and sweet peaceful reddest honest embroidered
    warm nose. Speak your mind! You are the sum of sum of sum of sum of hamster
    and charming happy angel and sweet cunning rich gentle hero and noble
    blossoming bold white amazing joy and cute honest white gentle tiny rich
    hamster. Speak thy mind! You are the sum of sum of sum of sum of good cat
    and cute honest pretty King and white tiny sweetest cunning cat and honest
    happy mighty delicious lovely town and rural fine honest lovely loving tiny
    town. Speak your mind! You are the sum of sum of sum of sum of sweet plum
    and mighty cunning rich joy and gentle little delicious charming kingdom
    and brave mighty rich white cute happiness and little sunny sunny white
    happy blossoming King. Speak thy mind! Thou art the sum of mighty summer's
    day and bold warm honest purse. Speak your mind! let us return to scene II.

                    Scene IV: Fizz.

Oberon:
     Thou art the sum of sum of sum of warm joy and noble sweetest plum and
    delicious cunning cute prompt cute cat and rural sweet trustworthy proud
    bold sweetest nose. Speak thine mind! Thou art the sum of sum of sum of
    town and cunning amazing peaceful town and brave fair sweetest handsome
    delicious King and old sweetest mighty cunning loving brave kingdom. Speak
    thy mind! Thou art the sum of sum of sum of sum of cunning pony and fair
    rich peaceful happiness and blossoming trustworthy sweet proud purse and
    happy healthy proud happy happy purse and cunning reddest embroidered
    cunning bold loving King. Speak thy mind! Thou art the sum of sum of sum of
    sum of delicious Heaven and healthy beautiful mighty kingdom and noble rich
    good tiny angel and fine brave cute beautiful handsome rose and proud
    healthy fine mighty amazing brave Heaven. Speak your mind! Thou art the sum
    of reddest happiness and charming sweet sweetest angel. Speak thy mind! We
    must proceed to scene II.

                    Scene V: Buzz.

Oberon:
     You are the sum of sum of trustworthy plum and good beautiful fine cunning
    proud hamster and gentle noble mighty honest golden pretty cat. Speak thy
    mind! You are the sum of sum of sum of sum of kingdom and white fine Lord
    and sweet good pretty beautiful plum and reddest handsome beautiful proud
    gentle rose and golden sweetest proud fair reddest gentle Lord. Speak thy
    mind! You are the sum of sum of sum of sum of peaceful King and cunning
    handsome handsome cat and amazing cunning happy beautiful nose and mighty
    old cute old peaceful sky and healthy amazing delicious mighty smooth
    embroidered town. Speak thine mind! Thou art the sum of sum of sum of sum
    of delicious Heaven and peaceful delicious charming flower and lovely bold
    happy old summer's day and gentle trustworthy noble good loving joy and
    delicious clearest handsome prompt healthy pretty purse. Speak thy mind!
    You are the sum of mighty happiness and sunny rural delicious joy. Speak
    your mind! We shall proceed to scene II.

                    Scene VI: End.

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots example programs should transpile hi.nfspl to SPL 1`] = `
"Don John and the bad hero.

Don John, the a variable.
Juliet, the b variable.

                    Act I: Main.

                    Scene I: Print hi.

[Enter Don John and Juliet]

Juliet:
     Thou art the sum of noble delicious handsome hamster and cunning good rich
    brave trustworthy gentle cat. Speak your mind!

Juliet:
     Thou art the sum of you and hamster. Speak thine mind!

Juliet:
     You are the sum of smooth flower and white bold blossoming cat. Speak
    thine mind!

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots example programs should transpile simplified reverse program 1`] = `
"Queen Elinor and the sweet fellow.

Queen Elinor, the stack variable.
Helena, the count variable.

                    Act I: Main.

                    Scene I: Init.

[Enter Queen Elinor and Helena]

Queen Elinor:
     You are nothing.

                    Scene II: Get input.

Helena:
     Open thy mind! Remember thyself!

Queen Elinor:
     You are the sum of you and town. Is you friendlier than nothing?

Helena:
     If so, Let us proceed to scene II. Recall your bluest blue leech!

                    Scene III: Print reversed.

Helena:
     Recall your old pretty town! Speak your mind!

Queen Elinor:
     Thou art the difference between you and Lord. Is you jollier than nothing?

Helena:
     If so, we shall return to scene III.

                    Scene IV: End.

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots language features should transpile arithmetic operations 1`] = `
"Christopher Sly and the oozing purse.

Christopher Sly, the calc variable.
Fenton, the result variable.

                    Act I: Main.

                    Scene I: Arithmetic test.

[Enter Christopher Sly and Fenton]

Christopher Sly:
     You are the sum of sum of sunny cat and trustworthy noble warm happiness
    and sum of summer's day and warm peaceful town. You are the difference
    between sum of mighty peaceful cat and sweetest pretty trustworthy golden
    purse and sum of sum of angel and lovely King and sunny mighty hamster. You
    are the product of sum of sunny King and cute happy rose and sum of sum of
    purse and warm purse and healthy beautiful Lord. You are the quotient
    between sum of sum of sum of Lord and clearest plum and cunning cunning sky
    and tiny clearest proud town and sum of pony and tiny Lord. You are the
    remainder of the quotient between sum of hero and rich smooth fair clearest
    sky and sum of cat and fair lovely rose.

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots language features should transpile basic I/O operations 1`] = `
"Ferdinand and the fat-kidneyed roman.

Ferdinand, the reader variable.
Agrippa, the writer variable.

                    Act I: Main.

                    Scene I: I o test.

[Enter Ferdinand and Agrippa]

Ferdinand:
     Open thy mind! Listen to thine heart!

Agrippa:
     Thou art the sum of purse and sweetest lovely tiny loving honest healthy
    summer's day. Speak thine mind! Thou art the sum of sum of proud cat and
    little prompt sweet pony and cunning fine cute good cute happiness. Open
    thy heart!

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots language features should transpile character literals 1`] = `
"Adriana and the little face.

Adriana, the char_writer variable.
Pantino, the output variable.

                    Act I: Main.

                    Scene I: Char test.

[Enter Adriana and Pantino]

Adriana:
     Thou art the sum of hero and charming sweet pretty tiny golden cunning
    angel. Thou art the sum of sum of sum of loving kingdom and pretty sweetest
    bold plum and proud gentle embroidered fair pony and peaceful mighty gentle
    beautiful tiny embroidered hero. Thou art the sum of good lovely
    embroidered amazing plum and loving gentle healthy golden lovely flower.
    Thou art the sum of sum of sum of flower and mighty rural cunning hero and
    honest pretty sweet blossoming flower and sweetest trustworthy white
    healthy brave hero.

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots language features should transpile comparison operations 1`] = `
"Sebastian and the stuffed mistletoe.

Sebastian, the left variable.
Orsino, the right variable.

                    Act I: Main.

                    Scene I: Comparison test.

[Enter Sebastian and Orsino]

Sebastian:
     Is you as handsome as Orsino? Is you fresher than Orsino? Is you worse
    than Orsino? Is you not as half-witted as Orsino? Is you not friendlier
    than Orsino? Is you not punier than Orsino?

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots language features should transpile control flow 1`] = `
"Solinus and the rotten beggar.

Solinus, the tester variable.
Lady Macbeth, the value variable.
Dogberry, the winner variable.
Leonato, the output variable.
Ophelia, the loser variable.

                    Act I: Main.

                    Scene I: Control test.

[Enter Solinus and Lady Macbeth]

Solinus:
     Is you as warm as sum of sum of cunning happiness and mighty smooth rural
    Heaven and old cute rich cunning sweetest sky? If so, Let us return to
    scene II. Is you nicer than nothing? If not, We must return to scene III.

                    Scene II: Success.

[Enter Dogberry and Leonato]

Dogberry:
     You are the sum of sum of sum of sum of summer's day and happy hero and
    golden bold hamster and cunning sweetest good proud flower and brave proud
    trustworthy fair mighty rural kingdom.

[Exeunt]

                    Scene III: Failure.

[Enter Ophelia and Leonato]

Ophelia:
     Thou art the sum of sum of little golden angel and lovely fine little town
    and handsome sweet fair trustworthy sweetest good town.

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots language features should transpile multiple acts and scenes 1`] = `
"Duncan and the huge summer's day.

Duncan, the a variable.
Solinus, the b variable.
Helena, the c variable.
Sebastian, the d variable.
Hector, the e variable.
Fortinbras, the f variable.

                    Act I: Act1.

                    Scene I: Scene1.

[Enter Duncan and Solinus]

Duncan:
     You are plum.

[Exeunt]

                    Scene II: Scene2.

[Enter Helena and Sebastian]

Helena:
     You are charming plum.

[Exeunt]

                    Act II: Act2.

                    Scene I: Scene1.

[Enter Hector and Fortinbras]

Hector:
     Thou art the sum of nose and rural hero.

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots language features should transpile positive numbers 1`] = `
"Lychorida and the clearest war.

Lychorida, the numbers variable.
Mistress Page, the output variable.

                    Act I: Main.

                    Scene I: Number test.

[Enter Lychorida and Mistress Page]

Lychorida:
     Thou art nothing. Thou art happiness. You are the sum of sum of fine
    kingdom and little amazing sweetest rose and brave handsome noble happy
    cute plum. Thou art the sum of sum of gentle warm rose and delicious loving
    beautiful proud peaceful Heaven and honest happy amazing cute charming
    brave cat.

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots language features should transpile stack operations 1`] = `
"Donalbain and the normal hog.

Donalbain, the pusher variable.
Egeus, the stack variable.

                    Act I: Main.

                    Scene I: Stack test.

[Enter Donalbain and Egeus]

Donalbain:
     Thou art the sum of sum of good nose and sweetest brave gentle Lord and
    amazing noble tiny blossoming charming hamster. Remember yourself! Remember
    me!

Egeus:
     Recall your smelly bluest Hell! Recall your fat fair grandmother!

[Exeunt]"
`;

exports[`NFSPL to SPL Snapshots language features should transpile variable references 1`] = `
"Goneril and the blossoming road.

Goneril, the a variable.
The Archbishop of Canterbury, the b variable.

                    Act I: Main.

                    Scene I: Variable test.

[Enter Goneril and The Archbishop of Canterbury]

Goneril:
     You are the sum of bold King and pretty brave amazing purse.

The Archbishop of Canterbury:
     Thou art the sum of Goneril and sum of happiness and warm loving Heaven.
    You are the product of Goneril and The Archbishop of Canterbury.

[Exeunt]"
`;
